src/plugins/content_management/server/event_stream/es/es_event_stream_client.ts

```diff
@@ -0,0 +1,197 @@
+    );
+
+    if (errors) {
+      throw new Error('Some events failed to be indexed.');
+    }
+  }
+
+  public async tail(limit: number = 100): Promise<EventStreamEvent[]> {
+    const esClient = await this.deps.esClient;
+    const res = await esClient.search<EsEventStreamEventDto>({
+      index: this.#names.dataStream,
+      query: {
+        match_all: {},
+      },
+      sort,
+      size: limit,
+    });
+    const events = res.hits.hits.map((hit) => dtoToEvent(hit._source!));
+
+    return events;
+  }
+
+  public async filter(
+    options: EventStreamClientFilterOptions
+  ): Promise<EventStreamClientFilterResult> {
+    const esClient = await this.deps.esClient;
+    const topLevelNodes: KueryNode[] = [];
+
+    if (options.subject && options.subject.length) {
+      topLevelNodes.push(
+        nodeBuilder.or(
+          options.subject.map(([type, id]) =>
+            !id
+              ? nodeBuilder.is('subjectType', type)
+              : nodeBuilder.and([
+                  nodeBuilder.is('subjectType', type),
+                  nodeBuilder.is('subjectId', id),
+                ])
+          )
+        )
+      );
+    }
+
+    if (options.object && options.object.length) {
+      topLevelNodes.push(
+        nodeBuilder.or(
+          options.object.map(([type, id]) =>
+            !id
+              ? nodeBuilder.is('objectType', type)
+              : nodeBuilder.and([
+                  nodeBuilder.is('objectType', type),
+                  nodeBuilder.is('objectId', id),
+                ])
+          )
+        )
+      );
+    }
+
+    if (options.predicate && options.predicate.length) {
+      topLevelNodes.push(
+        nodeBuilder.or(options.predicate.map((type) => nodeBuilder.is('predicate', type)))
+      );
+    }
+
+    if (options.transaction && options.transaction.length) {
+      topLevelNodes.push(
+        nodeBuilder.or(options.transaction.map((id) => nodeBuilder.is('txId', id)))
+      );
+    }
+
+    if (options.from) {
+      const from = new Date(options.from).toISOString();
+      const node = nodeBuilder.range('@timestamp', 'gte', from);
+
+      topLevelNodes.push(node);
+    }
+
+    if (options.to) {
+      const to = new Date(options.to).toISOString();
+      const node = nodeBuilder.range('@timestamp', 'lte', to);
+
+      topLevelNodes.push(node);
+    }
+
+    const query = toElasticsearchQuery(nodeBuilder.and(topLevelNodes));
+    const size = options.limit ?? 100;
+    const request: estypes.SearchRequest = {
+      index: this.#names.dataStream,
+      query,
+      sort,
+      size,
+    };
+
+    if (options.cursor) {
+      request.search_after = JSON.parse(options.cursor);
+    }
+
+    const res = await esClient.search<EsEventStreamEventDto>(request);
+    const events = res.hits.hits.map((hit) => dtoToEvent(hit._source!));
+    const lastHit = res.hits.hits[res.hits.hits.length - 1];
+
+    let cursor: string = '';
+
+    if (events.length >= size && lastHit && lastHit.sort) {
```

Comment:
__question\issue:__ I might be missing something, but how `events.length` can be greater than `size`? The `hits.total` can definitely be greater, but can ES return more hits than was requested?

Reply:
You are not missing anything, most likely ES will never return more results than requested. So `events.length == size` should be enough, in theory. But `events.length >= size` is safer.

Reply:
Yeah, I wasn't sure if you still wanted to return `cursor` in a case (albeit rare) when the `size` equals to `hits.total` (the last\only page). If it's intentional, then it'd probably make sense to use `track_total_hits: false` to be more performant (I expect this data stream might have a huge number of documents so every tiny bit will add up).

Reply:
`track_total_hits: false` is a good optimization suggesstion! Added it.